### PR TITLE
 Allow DMA on both halves of a split Serial<USART> separately.

### PR DESCRIPTION
**Pull request by bobmcwhirter**
Based upon and dependent upon https://github.com/stm32-rs/stm32f4xx-hal/pull/186

### CODE REVIEW

**Comment by therealprof** [comment]
@bobmcwhirter Mind rebasing this so we can have a look at the gist of this change?

**Comment by bobmcwhirter** [comment]
Will do shortly.

**Comment by bobmcwhirter** [comment]
rebased/repushed.

**Comment by thalesfragoso** [comment]
Hmm, seems like the first commit is undoing one of my last changes.

**Comment by bobmcwhirter** [comment]
I probably screwed up my rebase I guess. Will fix Monday.

**Comment by bobmcwhirter** [comment]
Hopefully this looks better.

File: src/serial.rs [hunk 1]
```diff
@@ -1442,6 +1453,27 @@ macro_rules! halUsartImpl {
...
+                            })
+                        }
+                        DmaConfig::TxRx => {
+                            usart.cr3.write(|w| {
+                                w
+                                    .dmar().enabled()
+                                    .dmat().enabled()
+                            })
+                        }
+                        _ => {}
```
Comment (thalesfragoso) [inline src/serial.rs:1474]: Maybe this should be `DmaConfig::None` instead ?

Comment (bobmcwhirter) [reply src/serial.rs:1474]: Ah, good point. I'll modify.

wrt hardware, yes, on an stm32f401re against my USART6.

**Pull request merged**